File: 2/2_cuatrimestre/PDOO/practicas/practica 2/deepSpaceRuby/GameUniverse.rb
require_relative './Dice.rb'
require_relative './GameStateController.rb'
require_relative './GameState'
require_relative 'GameUniverseToUI'
require_relative 'SpaceStation'
require_relative 'EnemyStarShip'

module Deepspace
    class GameUniverse

        def initialize()
            @WIN = 10 
            @gameState = GameStateController.new
            @dice = Dice.new
            @turns = 0      
            

            # no se si hay que inicializarlo asi
            @currentStationIndex = 0
            @currentStation = nil
            @spaceStations = nil
            @currentEnemy = nil
        end

        
        def combatGo (station,enemy) 

        end

        def combat

        end

        def discardHangar
            if (@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
                @spaceStations[@currentStationIndex].discardHangar()
            end
        end

        def discardShieldBooster(i)
            if (@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
                @spaceStations[@currentStationIndex].discardShieldBooster(i)
            end
        end
        
        def discardShieldBoosterInHangar(i)
            if (@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
                @spaceStations[@currentStationIndex].discardShieldBoosterInHangar(i)
            end
        end


        def discardWeapon(i)
            if (@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
                @spaceStations[@currentStationIndex].discardWeapon(i)
            end
        end


        def discardWeaponInHangar(i)
            if (@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
                @spaceStations[@currentStationIndex].discardWeaponInHangar(i)
            end
        end


        def getState
            return @gameState.state
        end



        def init(names)

        end


        def haveAWinner
            @currentStation.getNMedals >= @WIN
        end

        def mountShieldBooster(i)
            if (@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
                @spaceStations[@currentStationIndex].mountShieldBooster(i)
            end

        end

        def mountWeapon(i)
            if (@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
                @spaceStations[@currentStationIndex].mountWeapon(i)
            end
        end

        def nextTurn

        end

        def getUIversion
            return GameUniverseToUI.new(@currentStation, @currentEnemy)
        end

        def to_s
            return getUIversion.to_s
        end
    end 
end 

File: 2/2_cuatrimestre/PDOO/practicas/practica 2/deepSpace_2/src/deepspace/GameUniverse.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;


/**
 *
 * @author manbolq
 */
public class GameUniverse {
    private final int WIN = 10;
    
    private int currentStationIndex;
    private int turns;
    private Dice dice;
    private GameStateController gameState;
    private SpaceStation currentStation;
    private ArrayList<SpaceStation> spaceStations;
    private EnemyStarShip currentEnemy;
    
    
    /**************************************************************************/
    /*************************    CONSTRUCTORES   *****************************/
    /**************************************************************************/
    
    GameUniverse(){
        gameState = new GameStateController();
        dice = new Dice();
        turns = 0;
        
        // no se si hay que inicializarlo asi
        currentStationIndex = 0;
        currentStation = new SpaceStation("", new SuppliesPackage(0, 0, 0));
        spaceStations = new ArrayList<>();
        currentEnemy = new EnemyStarShip("", 0, 0, new Loot(0, 0, 0, 0, 0), new Damage(0, 0));
    }
    
    /**************************************************************************/
    /****************************    METODOS   ********************************/
    /**************************************************************************/
    
    CombatResult combat(SpaceStation station, EnemyStarShip enemy){
        throw new UnsupportedOperationException();
    }
    
    public CombatResult combat(){
        throw new UnsupportedOperationException();
    }
    
    public void discardHangar(){
        if ((gameState.getState() == GameState.INIT) || (gameState.getState() == GameState.AFTERCOMBAT))
            spaceStations.get(currentStationIndex).discardHangar();
    }
    
    public void discardShieldBooster(int i){
        if ((gameState.getState() == GameState.INIT) || (gameState.getState() == GameState.AFTERCOMBAT))
            spaceStations.get(currentStationIndex).discardShieldBooster(i);
    }
    
    public void discardShieldBoosterInHangar(int i){
         if ((gameState.getState() == GameState.INIT) || (gameState.getState() == GameState.AFTERCOMBAT))
            spaceStations.get(currentStationIndex).discardShieldBoosterInHangar(i);
    }
    
    public void discardWeapon(int i){
         if ((gameState.getState() == GameState.INIT) || (gameState.getState() == GameState.AFTERCOMBAT))
            spaceStations.get(currentStationIndex).discardWeapon(i);
    }
    
    public void discardWeaponInHangar(int i){
         if ((gameState.getState() == GameState.INIT) || (gameState.getState() == GameState.AFTERCOMBAT))
            spaceStations.get(currentStationIndex).discardWeaponInHangar(i);
    }
    
    public GameState getState(){
        return gameState.getState();
    }
    
    public GameUniverseToUI getUIversion(){
        return new GameUniverseToUI(currentStation, currentEnemy);
    }
    
    public boolean haveAWinner(){
        return currentStation.getNMedals() >= WIN;
    }
    
    public void init(ArrayList<String> names){
        throw new UnsupportedOperationException();
    }
    
    public void mountShieldBooster(int i){
        if ((gameState.getState() == GameState.INIT) || (gameState.getState() == GameState.AFTERCOMBAT))
            spaceStations.get(currentStationIndex).mountShieldBooster(i);
    }
    
    public void mountWeapon(int i){
        if ((gameState.getState() == GameState.INIT) || (gameState.getState() == GameState.AFTERCOMBAT))
            spaceStations.get(currentStationIndex).mountWeapon(i);
    }
    
    public boolean nextTurn(){
        throw new UnsupportedOperationException();
    }
    
    @Override
    public String toString(){
        return getUIversion().toString();
    }
    
}
